### DISCUSSION TITLE
Read data from submitted forms and display it to the user that submitted the form

### DISCUSSION THREAD

**Discussion by PanickingLynx** [sulu/SuluFormBundle#334] [Q&A] [answered]
Hello community!
I just got started with the Sulu Form Bundle and would like to know if there is any straightforward way of enabling the user to view the things they've submitted to my form. Also, it would be great if the form data could also include the name of the user that submitted the request.

So if it is something like a support request form, that they can see all their submitted issues and the progress associated with them. 

Best regards
_Panic_

**Comment by PanickingLynx** [comment]
So I Kind of found a workaround.
By calling the dynamic Form API directly (I did it VERY cheaply with jQuery Ajax), and adding the API Route as a firewall exception (ROLE_USER since the reading happens behind another firewall security Zone than Admin) in `security.yml` . I was successfully abled to retrieve the Form Data in JSON. I will Mark this as solved for now and edit with a more comprehensive answer in the next days.

**Comment by PanickingLynx** [comment]
Okay, now the way I solve this is **NOT** the pretty way of doing things, but this is just a POC and would need to carried over to Symfony.

I went ahead and added `- { path: ^/admin/api/form/dynamics, roles: ROLE_SULU_MIETER }` to `/config/packages/security.yaml`.
This allows access to the API endpoint provided by the bundle. (ROLE_SULU_MIETER is my usergroup with access rights to the control panel for tenants in my webapp)

Then to retrieve all data on my frontend, I wrote up this pretty hacky piece of jQuery:
```js
        $(document).ready(function(){
            $.get( "/admin/api/form/dynamics", { 
                "page": 1,
                "locale": "en",
                "form": 2,
                "sortBy": "created",
                "sortOrder": "desc",
                "limit": 10,
                "fields": ["id","text","radioButtons","textarea","attachment","text1","text2","created"],
                "flat": true 
            },
            function( data ) {
                var ResponseData = data["_embedded"].dynamic_forms;
                ResponseData.forEach(
                    function(element){
                        var markup = "<tr><td>" + element.text + "</td>" + "<td>" + element.radioButtons + "</td>" + "<td>" + element.textarea + "</td>" + "<td>" + element.created + "</tr></td>";
                        if (element.text1 == '{{ app.user.getUsername() }}'){
                            $('#Reports tbody').append(markup);
                        }
                    }
                );
            });
        });
```

Again, **JUST A POC**, I know this is very insecure and very very bad practice, **SO PLEASE DO NOT CARRY THIS OVER TO YOUR OWN PROJECT!**

This allows me to retrieve the data submitted to the forms away from the `/admin` section.
I will update this thread as I explore the possibilities to make this actually usable in a real application.
This example also only works with dynamic forms and only if you know the object returned by the API in the form wizard. 

I am sadly not proficient enough with OOPHP and Symfony to pull it off cleanly.

Hope someone finds this atleast a little helpful.

Best regards
_Panic_

**Comment by alexander-schranz** [comment, marked as answer]
The sulu form bundle makes a redirect after the form is submitted. This is done to avoid duplicated form submits if a user does a page refresh. That make all data which is submitted to be lost on the target `?send=true`.

Still it is possible that you listen the the [FormSavePostEvent](https://github.com/sulu/SuluFormBundle/blob/4b52fdfb623e126e43b38fa0acfce3af5b0982a5/Form/Handler.php#L137-L143). Then write this data into the session of the user or even better into the flashbag of the session:  https://symfony.com/doc/current/components/http_foundation/sessions.html

You create a new controller which read that data from the current user session / flashbag and return it as a `JSON`. If you are on the `?send=true` page your JS need to request that new endpoint you created it and show so the submitted data.

By reading it over via JS over ajax you make sure that also the `?send=true` page can be cached by Sulu Http Cache and only the required user specific data need to be requested from the session of the user. 

You should not expose any admin api to the public as that will leak private information of other users.

**Reply by PanickingLynx** [reply]
Yeah, I sadly lack the proficiency to just build my own controller, which is probably the issue at hand here.

**Reply by alexander-schranz** [reply]
Sulu doesn't hide that it is build on top of Symfony framework, its also its advantages as if Sulu does not provide something you can at the end just think how you would solve a problem with Symfony and do it that way.

**Reply by PanickingLynx** [reply]
Yeah I know. I was put on the project without having any knowledge of Symfony let alone PHP being told to "Teach it myself as I Go along" which is why I asked so many questions here. I will seize this for now since I don't want to be a bother, but thank you for your patience.

**Reply by alexander-schranz** [reply]
Symfony has a really [good documentation](https://symfony.com/doc/current/index.html) there is also a quick tour documentation to get into it: https://symfony.com/doc/current/quick_tour/the_big_picture.html. Still feel free to ask we will do our best to answer them.
